src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -1195,22 +1192,56 @@ void addSubpatternsForTuple(ImmutableArray<TypeWithAnnotations> elementTypes)
             DiagnosticBag diagnostics)
         {
             BoundExpression value = BindExpressionForPattern(inputType, node.Expression, hasErrors, diagnostics, out var constantValueOpt, out _);
+            BinaryOperatorKind operation = TokenKindToBinaryOperatorKind(node.OperatorToken.Kind());
+            BinaryOperatorKind opType = BinaryOperatorType(value.Type!.SpecialType);
+            switch (opType)
+            {
+                case BinaryOperatorKind.Float:
+                case BinaryOperatorKind.Double:
+                    if (!hasErrors && constantValueOpt != null && double.IsNaN(constantValueOpt.DoubleValue))
+                    {
+                        diagnostics.Add(ErrorCode.ERR_RelationalPatternWithNaN, node.Expression.Location);
+                        hasErrors = true;
+                    }
+                    break;
+                case 0:
+                    if (!hasErrors)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_UnsupportedTypeForRelationalPattern, node.Location, value.Type.ToDisplayString());
+                        hasErrors = true;
+                    }
+                    break;
+            }
 
-            // PROTOTYPE(ngafter): TODO: check that the operator exists for that type (e.g. error for string < "")
-            BinaryOperatorKind kind = TokenKindToBinaryOperatorKind(node.OperatorToken.Kind());
-            return new BoundRelationalPattern(node, kind, value, constantValueOpt ?? ConstantValue.Bad, inputType, hasErrors);
-        }
+            if (constantValueOpt is null) hasErrors = true;
+            return new BoundRelationalPattern(node, operation | opType, value, constantValueOpt ?? ConstantValue.Bad, inputType, hasErrors);
 
-        private BinaryOperatorKind TokenKindToBinaryOperatorKind(SyntaxKind kind) => kind switch
-        {
-            SyntaxKind.LessThanEqualsToken => BinaryOperatorKind.LessThanOrEqual,
-            SyntaxKind.LessThanToken => BinaryOperatorKind.LessThan,
-            SyntaxKind.GreaterThanToken => BinaryOperatorKind.GreaterThan,
-            SyntaxKind.GreaterThanEqualsToken => BinaryOperatorKind.GreaterThanOrEqual,
-            // The following occur in error recovery scenarios
-            SyntaxKind.ExclamationEqualsToken => BinaryOperatorKind.NotEqual,
-            _ => BinaryOperatorKind.Equal,
-        };
+            static BinaryOperatorKind TokenKindToBinaryOperatorKind(SyntaxKind kind) => kind switch
+            {
+                SyntaxKind.LessThanEqualsToken => BinaryOperatorKind.LessThanOrEqual,
+                SyntaxKind.LessThanToken => BinaryOperatorKind.LessThan,
+                SyntaxKind.GreaterThanToken => BinaryOperatorKind.GreaterThan,
+                SyntaxKind.GreaterThanEqualsToken => BinaryOperatorKind.GreaterThanOrEqual,
+                // The following occurs in error recovery scenarios
+                _ => BinaryOperatorKind.Equal,
```

Comment:
>Equal [](start = 40, length = 5)

nit: This is existing code, but I wonder why return `Equal` here, rather than `None` or some other value more appropriate to error scenarios?

Reply:
Because I want this to flow through the decision dag builder and the value sets without requiring additional accommodation.  This is not used as a flag to the caller that there is an error condition; if that were the case an error indicator would be appropriate.

---
In reply to: [380986697](https://github.com/dotnet/roslyn/pull/41503#discussion_r380986697) [](ancestors = 380986697)